[WIP] Add android target support

omkar-tenkale opened the pull request:
Only the basic triangle example is working as of now

Few more things are to be done
- Make it run all examples successfully

<img width="267" height="471" alt="image" src="https://github.com/user-attachments/assets/9d121b92-8760-4093-8c5d-483e0142a6a7" />

omkar-tenkale commented on core/src/androidMain/kotlin/com/littlekt/graphics/webgpu/webgpu.android.kt:
```diff
@@ -0,0 +1,112 @@
+}
+
+actual fun getNativeWebGPUTextureByteSize(texture: WGPUTexture): Long {
+    return Native.POINTER_SIZE.toLong()
```
This is likely wrong. But I couldn't get the texture size on Android target because of the lightweight pointer with just the reference

omkar-tenkale commented on core/src/androidMain/kotlin/com/littlekt/AndroidGraphics.kt:
```diff
@@ -0,0 +1,83 @@
+        val nativeSurface = wgpu.getSurfaceFromAndroidWindow(window) ?: error("Can't create Surface")
+        surface = Surface(nativeSurface.handler)
+        if (::adapter.isInitialized.not()) {
+            adapter = Adapter((wgpu.requestAdapter(nativeSurface) ?: error("Can't create Adapter")).handler)
```
fixed in latest commit
https://github.com/littlektframework/littlekt/pull/313/commits/9cd14959e23b5889099a922451bb7d86ca7a0c04

omkar-tenkale commented on core/src/androidMain/kotlin/com/littlekt/graphics/webgpu/webgpu.android.kt:
```diff
@@ -0,0 +1,112 @@
+package com.littlekt.graphics.webgpu
```
@LeHaine — over to you from here, I’ve taken it as far as I could 🙂

LeHaine replied:
I can check this out sometime this weekend when I have some free time. Nice job, thus far!

omkar-tenkale commented:
This example is not working on android but works on jvm and web. im assuming issue with Buffer impl

​

```

package com.littlekt.examples

​

import com.littlekt.Context

import com.littlekt.ContextListener

import com.littlekt.file.FloatBuffer

import com.littlekt.file.ShortBuffer

import com.littlekt.graphics.Color

import com.littlekt.graphics.webgpu.*

​

/**

* A test to isolate texture writeTexture issues by using hardcoded data instead of VFS loading.

*/

class HardcodedTextureTest(context: Context) : ContextListener(context) {

// language=wgsl

private val textureShader =

"""

struct VertexOutput {

@location(0) uv: vec2<f32>,

@builtin(position) position: vec4<f32>,

};

@vertex

fn vs_main(

@location(0) pos: vec2<f32>,

@location(1) uvs: vec2<f32>) -> VertexOutput {

var output: VertexOutput;

output.position = vec4<f32>(pos.x, pos.y, 0, 1);

output.uv = uvs;

return output;

}

@group(0) @binding(0)

var my_texture: texture_2d<f32>;

@group(0) @binding(1)

var my_sampler: sampler;

@fragment

fn fs_main(in: VertexOutput) -> @location(0) vec4<f32> {

return textureSample(my_texture, my_sampler, in.uv);

}

"""

.trimIndent()

​

override suspend fun Context.start() {

addStatsHandler()

logger.info { "HardcodedTextureTest.start() begin" }

val vertices =

FloatBuffer(

floatArrayOf(

-0.5f,

-0.5f,

0f,

1f,

-0.5f,

0.5f,

0f,

0f,

0.5f,

0.5f,

1f,

0f,

0.5f,

-0.5f,

1f,

1f,

)

)

val indices = ShortBuffer(shortArrayOf(0, 1, 2, 0, 2, 3))

logger.debug {

val vCount = vertices.capacity

val vBytes = vCount * Float.SIZE_BYTES

val iCount = indices.capacity

val iBytes = iCount * Short.SIZE_BYTES

"buffers prepared: vertices count=$vCount bytes=$vBytes, indices count=$iCount bytes=$iBytes"

}

​

// Create a simple 2x2 RGBA texture with hardcoded data

val textureWidth = 2

val textureHeight = 2

val textureData = byteArrayOf(

// Pixel 0,0: Red

255.toByte(), 0, 0, 255.toByte(),

// Pixel 1,0: Green

0, 255.toByte(), 0, 255.toByte(),

// Pixel 0,1: Blue

0, 0, 255.toByte(), 255.toByte(),

// Pixel 1,1: White

255.toByte(), 255.toByte(), 255.toByte(), 255.toByte()

)

​

val device = graphics.device

logger.debug {

val limits = device.limits

"device limits: maxBufferSize=${limits.maxBufferSize} maxVertexBuffers=${limits.maxVertexBuffers} maxVertexAttributes=${limits.maxVertexAttributes}"

}

logger.info { "creating GPU buffers via createBuffer + writeBuffer..." }

val queue = device.queue

val vboSize = vertices.capacity.toLong() * Float.SIZE_BYTES

logger.debug { "createBuffer label=vbo usage=VERTEX|COPY_DST bytes=$vboSize mapped=false" }

val vbo = device.createBuffer(BufferDescriptor("vbo", vboSize, BufferUsage.VERTEX or BufferUsage.COPY_DST, false))

logger.debug { "queue.writeBuffer vbo elements=${vertices.capacity}" }

queue.writeBuffer(vbo, vertices, 0, 0, vertices.capacity.toLong())

logger.debug { "vbo ready size=${vbo.size}" }

val iboSize = indices.capacity.toLong() * Short.SIZE_BYTES

logger.debug { "createBuffer label=ibo usage=INDEX|COPY_DST bytes=$iboSize mapped=false" }

val ibo = device.createBuffer(BufferDescriptor("ibo", iboSize, BufferUsage.INDEX or BufferUsage.COPY_DST, false))

logger.debug { "queue.writeBuffer ibo elements=${indices.capacity}" }

queue.writeBuffer(ibo, indices, 0, 0, indices.capacity.toLong())

logger.debug { "ibo ready size=${ibo.size}" }

logger.info { "creating shader module..." }

val shader = device.createShaderModule(textureShader)

logger.debug { "shader module created" }

val surfaceCapabilities = graphics.surfaceCapabilities

val preferredFormat = graphics.preferredFormat

logger.debug {

"graphics: preferredFormat=$preferredFormat srgb=${preferredFormat.srgb} surfaceCaps(formats=${surfaceCapabilities.formats.size}, alphaModes=${surfaceCapabilities.alphaModes.size})"

}

val texture =

device.createTexture(

TextureDescriptor(

Extent3D(textureWidth, textureHeight, 1),

1,

1,

TextureDimension.D2,

if (preferredFormat.srgb) TextureFormat.RGBA8_UNORM_SRGB

else TextureFormat.RGBA8_UNORM,

TextureUsage.COPY_DST or TextureUsage.TEXTURE,

)

)

logger.debug { "texture created sizeBytes=${texture} dims=${textureWidth}x${textureHeight}" }

​

// queue already obtained above

// Test writeTexture with hardcoded ByteArray - this will tell us if the issue is in VFS or WebGPU

logger.info { "queue.writeTexture begin" }

logger.debug {

val bytesPerRow = textureWidth * 4

val rowsPerImage = textureHeight

"writeTexture dataBytes=${textureData.size} layout(bytesPerRow=$bytesPerRow, rowsPerImage=$rowsPerImage) copySize=${textureWidth}x${textureHeight}x1"

}

queue.writeTexture(

data = textureData,

destination = TextureCopyView(texture),

layout = TextureDataLayout(textureWidth * 4, textureHeight),

copySize = Extent3D(textureWidth, textureHeight, 1),

)

logger.debug { "queue.writeTexture end" }

​

val sampler = device.createSampler(SamplerDescriptor())

val textureView = texture.createView()

logger.debug { "sampler and textureView created" }

val bindGroupLayout =

device.createBindGroupLayout(

BindGroupLayoutDescriptor(

listOf(

BindGroupLayoutEntry(0, ShaderStage.FRAGMENT, TextureBindingLayout()),

BindGroupLayoutEntry(1, ShaderStage.FRAGMENT, SamplerBindingLayout()),

)

)

)

logger.debug { "bindGroupLayout created" }

val bindGroup =

device.createBindGroup(

BindGroupDescriptor(

bindGroupLayout,

listOf(

BindGroupEntry(0, textureView),

BindGroupEntry(1, sampler),

),

)

)

logger.debug { "bindGroup created" }

​

val pipelineLayout = device.createPipelineLayout(PipelineLayoutDescriptor(listOf(bindGroupLayout)))

logger.debug { "pipelineLayout created" }

val renderPipelineDesc =

RenderPipelineDescriptor(

layout = pipelineLayout,

vertex =

VertexState(

module = shader,

entryPoint = "vs_main",

buffers =

listOf(

WebGPUVertexBufferLayout(

arrayStride = 4L * Float.SIZE_BYTES,

stepMode = VertexStepMode.VERTEX,

attributes =

listOf(

WebGPUVertexAttribute(

VertexFormat.FLOAT32x2,

offset = 0,

shaderLocation = 0,

),

WebGPUVertexAttribute(

VertexFormat.FLOAT32x2,

offset = 2L * Float.SIZE_BYTES,

shaderLocation = 1,

),

),

)

),

),

fragment =

FragmentState(

module = shader,

entryPoint = "fs_main",

target =

ColorTargetState(

format = preferredFormat,

blendState = BlendState.Opaque,

writeMask = ColorWriteMask.ALL,

),

),

primitive = PrimitiveState(topology = PrimitiveTopology.TRIANGLE_LIST),

depthStencil = null,

multisample =

MultisampleState(count = 1, mask = 0xFFFFFFF, alphaToCoverageEnabled = false),

)

val renderPipeline = device.createRenderPipeline(renderPipelineDesc)

logger.debug { "renderPipeline created" }

graphics.configureSurface(

TextureUsage.RENDER_ATTACHMENT,

preferredFormat,

PresentMode.FIFO,

surfaceCapabilities.alphaModes[0],

)

logger.debug { "surface configured ${graphics.width}x${graphics.height} format=$preferredFormat" }

​

onUpdate {

val surfaceTexture = graphics.surface.getCurrentTexture()

when (val status = surfaceTexture.status) {

TextureStatus.SUCCESS -> {

// all good, could check for `surfaceTexture.suboptimal` here.

}

TextureStatus.TIMEOUT,

TextureStatus.OUTDATED,

TextureStatus.LOST -> {

surfaceTexture.texture?.release()

graphics.configureSurface(

TextureUsage.RENDER_ATTACHMENT,

preferredFormat,

PresentMode.FIFO,

surfaceCapabilities.alphaModes[0],

)

logger.info { "getCurrentTexture status=$status" }

return@onUpdate

}

else -> {

// fatal

logger.fatal { "getCurrentTexture status=$status" }

close()

}

}

val texture = checkNotNull(surfaceTexture.texture)

val frame = texture.createView()

​

val commandEncoder = device.createCommandEncoder()

val renderPassEncoder =

commandEncoder.beginRenderPass(

desc =

RenderPassDescriptor(

listOf(

RenderPassColorAttachmentDescriptor(

view = frame,

loadOp = LoadOp.CLEAR,

storeOp = StoreOp.STORE,

clearColor = Color.CLEAR,

)

)

)

)

renderPassEncoder.setPipeline(renderPipeline)

renderPassEncoder.setBindGroup(0, bindGroup)

renderPassEncoder.setVertexBuffer(0, vbo)

renderPassEncoder.setIndexBuffer(ibo, IndexFormat.UINT16)

renderPassEncoder.drawIndexed(6, 1)

renderPassEncoder.end()

renderPassEncoder.release()

​

val commandBuffer = commandEncoder.finish()

​

queue.submit(commandBuffer)

logger.trace { "frame submitted" }

graphics.surface.present()

​

commandBuffer.release()

commandEncoder.release()

frame.release()

texture.release()

}

​

onRelease {

logger.info { "HardcodedTextureTest.onRelease()" }

bindGroup.release()

bindGroupLayout.release()

textureView.release()

sampler.release()

texture.release()

renderPipeline.release()

pipelineLayout.release()

shader.release()

vbo.release()

ibo.release()

}

}

}

​

```

LeHaine commented:
> any updates @LeHaine ?

Disregard previous comment. Do we have an error?

omkar-tenkale commented:
@LeHaine bumping just in case ^

Commit:
Add android target support

Commit:
Android: align JVM 25 and harden WGPU logging

Raise Android compiler/Java targets to 25, bump android-native-helper, and guard WGPU logging init to avoid runtime FFM failures.

omkar-tenkale commented:
hey @LeHaine can you take over this please? I'd love to see littlekt supporting android but seems wgpu is the only blocker currently. Its failing on my device but not failing on @ygdrasil-io 's 
Is it failing on yours?
https://github.com/wgpu4k/wgpu4k/issues/204#issuecomment-3767669838

LeHaine commented:
> hey @LeHaine can you take over this please? I'd love to see littlekt supporting android but seems wgpu is the only blocker currently. Its failing on my device but not failing on @ygdrasil-io 's Is it failing on yours? [wgpu4k/wgpu4k#204 (comment)](https://github.com/wgpu4k/wgpu4k/issues/204#issuecomment-3767669838)

Hello - I can test it out on my device and report back.

LeHaine commented:
@omkar-tenkale looks like I get the same error as you are seeing. Using real device as well. Will try to dig a bit further.


```
java.lang.InstantiationError: java.lang.foreign.MemorySegment
         at java.lang.foreign.SegmentAllocator.allocate(SegmentAllocator.kt:11)
         at ffi.MemoryAllocator.bufferOf(MemoryAllocator.android.kt:24)
         at ffi.MemoryAllocator.bufferOfAddress(MemoryAllocator.android.kt:33)
         at io.ygdrasil.webgpu.WGPU.requestAdapter(WGPU.kt:64)
         at io.ygdrasil.webgpu.WGPU.requestAdapter$default(WGPU.kt:39)
         at com.littlekt.AndroidGraphics.handleSurfaceCreated(AndroidGraphics.kt:63)
         at com.littlekt.AndroidContext$start$2.invokeSuspend(AndroidContext.kt:60)
         at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:34)
         at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:100)
         at kotlinx.coroutines.EventLoop.processUnconfinedEvent(EventLoop.common.kt:65)
         at kotlinx.coroutines.internal.DispatchedContinuationKt.resumeCancellableWith(DispatchedContinuation.kt:383)
         at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable(Cancellable.kt:26)
         at kotlinx.coroutines.CoroutineStart.invoke(CoroutineStart.kt:358)
         at kotlinx.coroutines.AbstractCoroutine.start(AbstractCoroutine.kt:134)
         at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch(Builders.common.kt:53)
         at kotlinx.coroutines.BuildersKt.launch(Unknown Source:1)
         at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch$default(Builders.common.kt:44)
         at kotlinx.coroutines.BuildersKt.launch$default(Unknown Source:1)
         at com.littlekt.LittleKtSurfaceView.surfaceCreated(LittleKtSurfaceView.kt:89)
```

LeHaine commented:
From what I can tell, it's attempting to use the JDK `MemorySegment` implementation which is an interface instead of using `wgpu4k-native` implementation, hence the instantiate error. This might just be a dependency / build issue.
